src/main/kotlin/org/rust/lang/core/completion/RsCommonCompletionProvider.kt

```diff
@@ -104,23 +104,26 @@ object RsCommonCompletionProvider : RsCompletionProvider() {
             is RsMetaItem -> return
             else -> {
                 val lookup = ImplLookup.relativeTo(element)
+                var filtered: RsResolveProcessor = filterPathCompletionVariantsByTraitBounds(
+                    addProcessedPathName(processor, processedPathNames),
+                    lookup
+                )
+                // Filters are applied in reverse order (the last filter is applied first)
+                val filters = listOf(
+                    ::filterCompletionVariantsByVisibility,
+                    ::filterAssocTypes,
+                    ::filterVisRestrictionPaths,
+                    ::filterTraitRefPaths
```

Comment:
Looks like re-inventing #2920 step by step... =D

Reply:
Uhh :sweat_smile: I have to say that your proposal seems much more readable :) I have no idea about how "debuggable" it is though, sometimes it must be painful to step through the resolve code.